### PR TITLE
 feat: color: add HSV methods to slint

### PR DESCRIPTION
**Pull request by flukejones**
TODO:
- docs
- finalize naming (brightness or value for `v` component
- methods or properties?

### CODE REVIEW

**Comment by flukejones** [comment]
Ref issue https://github.com/slint-ui/slint/issues/911

**Comment by flukejones** [comment]
I had some extra time to play more with this and tried out making it member-field access instead of function - I don't think there's any kind of easy way for me to do that without a new `Type::ColorHsva` which I don't want to do.

I also didn't want to add more fields on the Color type as that would affect the struct size.

Unless y'all think that adding a new slint type is a good thing to do? If so, I'd best leave it to the pros, otherwise I will complete this PR soon.

**Review by ogoffart — COMMENTED** [review]
Thank you so much for your work.

I'm a bit busy this week, so I won't comment much on the implementation at this point. But I just want to get the API right.

As is, this PR adds:

- Functions on the Slint color and brush type:
  * `hue()->float` (return value between 0.0 and 360.0) 
  * `saturation()->float` (return value between 0.0 and 255.0)
  * `brightness()->float` (return value between 0.0 and 255.0)
- Functions in the Colors namespace:
  * `hsv` and `hsva` which are alias to eachother. It seems that the hue is a float between 0 and 360, when saturation, value, and alpha are flaot between 0.0 and 1.0
- Rust API on the Color struct
    ```rust
    /// Converts this color to a tuple of `(hue, saturation, value)`
    pub fn to_hsv(&mut self) -> (f32, f32, f32) 

    /// Converts this color to a tuple of `(hue, saturation, value, alpha)`
    pub fn to_hsva(&mut self) -> (f32, f32, f32, f32)

    /// Construct a color from the hue, saturation, and value HSV color parameters. The alpha
    /// channel will have the value 1.0.
    pub fn from_hsv(hue: f32, saturation: f32, value: f32) -> Self

    /// Construct a color from the hue, saturation, and value HSV color parameters.
    pub fn from_hsva(hue: f32, saturation: f32, value: f32, alpha: f32) -> Self 
    
    
    /// Returns the hue channel of the color as f32 in degrees 0..PI.
    pub fn hue(mut self) -> f32

    /// Returns the saturation of the color as u8 in the range 0..255.
    pub fn saturation(mut self) -> f32 

    /// Returns the brightness of the color as u8 in the range 0..255.
    pub fn brightness(mut self) -> f32
    ```    
- Not included in this patch, but something will be needed in the C++ API as well

----

One thing to note is that the compiler can do transformations so that the Rust API and the slint API don't have to be the same.
For example, wether or not `hue` is a field or a function in Slint is just a matter of having an Expression::Call or not in the lookup.rs 
(for example, `red` is a field in slint and is calling BuiltinFunction::ColorRgbaStruct in the generated code before accessing a field)
Also you don't need to add a `Type::ColorHsva`, you can just use anonymous types such as `{a: float, h: angle, s: float, v: float}` which 
is represented in Rust as a tuple `(f32, f32, f32, f32)` and in C++ `std::tuple<float, float, float, float>`. This can be an internal detail 
in rust.rs and cpp.rs which is not part of the C++ and Rust API.

Given that, we have much freedom to do anything.


So my comments/questions on the API
 1. for the hue in Slint: should we use the `angle` type or the `float` type or the `int` type? I think i'd go with float since one probably want to do some math on it and not do `backround.hue/1deg` all the time.
 2. for the brightness and saturation, it is sometimes between 0 and 1 and sometimes between 0 and 255, this is not consistent. I'd use the 0 to 1 range consistently. Although we must also consider what the rgb or alpha values are.
 3. I have commented on the Slint API in this comment https://github.com/slint-ui/slint/issues/911#issuecomment-1980970787 , I think it should either be a field: `background.hue` or a function that return an anonymous type: `background.to-hsv().h` the later would allow to dinstinguish the saturation between the HSV and HSL color space.
 4. Do we need both Colors.hsv and Colors.hsva? rgb and rgba both exist for consistency with CSS, but there is no hsv in CSS. (there is hsl() and a bunch of other) I'd go with only the former.
 5. For the Rust API, we probably don't need both to_hsv and the getter on each component. I wonder if the best would not be to make the `HsvaColor` type public and have conversion between these.  Otherwise i think a tuple is not that bad. But the range needs to be well documented.  (and if we go for the 0..255, it should probably be u8)
 6. Same for the C++ API
 
@tronical: opinions on the API?

File: docs/reference/src/language/syntax/types.md [hunk 1]
```diff
@@ -81,6 +81,11 @@ All properties are in the range 0-255.
 
 All colors and brushes define the following methods:
 
+- **`linear-mix(other: brush, factor: float) -> brush`**
```
[the remark this answers is not in the document]
Comment (flukejones) [reply docs/reference/src/language/syntax/types.md:84]: Looks like a leftover from a rebase issue I had. Removed now.

**Comment by flukejones** [comment]
>     1. for the hue in Slint: should we use the `angle` type or the `float` type or the `int` type? I think i'd go with float since one probably want to do some math on it and not do `backround.hue/1deg` all the time.

I agree with this. The reason I need hsv myself is because I want to do math on it.

>     2. for the brightness and saturation, it is sometimes between 0 and 1 and sometimes between 0 and 255, this is not consistent.

Documentation mistake by me. Corrected now.

>     3. I have commented on the Slint API in this comment [Add properties to the color type to get HSV component #911 (comment)](https://github.com/slint-ui/slint/issues/911#issuecomment-1980970787) , I think it should either be a field: `background.hue` or a function that return an anonymous type: `background.to-hsv().h` the later would allow to dinstinguish the saturation between the HSV and HSL color space.

One of my concerns was the addition of extra colourspaces. So I agree with later option.

>     4. Do we need both Colors.hsv and Colors.hsva? rgb and rgba both exist for consistency with CSS, but there is no hsv in CSS. (there is hsl() and a bunch of other) I'd go with only the former.

No alpha?

>     5. For the Rust API, we probably don't need both to_hsv and the getter on each component. I wonder if the best would not be to make the `HsvaColor` type public and have conversion between these.  Otherwise i think a tuple is not that bad. But the range needs to be well documented.  (and if we go for the 0..255, it should probably be u8)

I would prefer making `HsvaColor` public myself. It's already there and opens the door to more types of colour handling.

**Comment by flukejones** [comment]
I'm afraid I will have to leave this as is now. I have too many tasks at work. I converted to struct member access with the goal of getting to `some_colour.hsva().hue` but how to accomplish this is not clear to me and I don't have enough time to invest.

@tronical or @ogoffart if one of you could please take over?

**Comment by flukejones** [comment]
Please also note I have not touched the C++ bindings - this is not one of my strong suits

File: api/cpp/include/slint_color.h [hunk 2]
```diff
@@ -37,6 +37,19 @@ struct RgbaColor
     RgbaColor(const Color &col);
 };
 
+/// HsvaColor stores the hue, saturation, value and alpha components of a color.
+struct HsvaColor
+{
+    /// The hue component in degrees between 0 and 360
```
Comment (tronical) [inline api/cpp/include/slint_color.h:43]:
```suggestion
    /// The hue component in degrees between 0 and 360.
```

File: api/cpp/include/slint_color.h [hunk 3]
```diff
@@ -37,6 +37,19 @@ struct RgbaColor
...
+    float hue;
+    /// The saturation component, between 0 and 1
+    float saturation;
+    /// The value component, between 0 and 1
```
Comment (tronical) [inline api/cpp/include/slint_color.h:47]:
```suggestion
    /// The value component, between 0 and 1.
```

File: api/cpp/include/slint_color.h [hunk 4]
```diff
@@ -37,6 +37,19 @@ struct RgbaColor
...
+    float value;
+    /// The alpha component, between 0 and 1
```
Comment (tronical) [inline api/cpp/include/slint_color.h:49]:
```suggestion
    /// The alpha component, between 0 and 1.
```

File: api/cpp/include/slint_color.h [hunk 5]
```diff
@@ -117,22 +130,40 @@ class Color
...
     /// Converts this color to an RgbaColor struct for easy destructuring.
-    inline RgbaColor<uint8_t> to_argb_uint() const;
+    [[nodiscard]] inline RgbaColor<uint8_t> to_argb_uint() const;
 
     /// Converts this color to an RgbaColor struct for easy destructuring.
-    inline RgbaColor<float> to_argb_float() const;
+    [[nodiscard]] inline RgbaColor<float> to_argb_float() const;
+
+    /// Construct a color from the HSV color space component.
+    /// The hue is comprised between 0 and 360, and the other parameter between 0 and 1
```
Comment (tronical) [inline api/cpp/include/slint_color.h:139]: Use simpler english:

```suggestion
    /// The hue is expected to be in the range between 0 and 360, and the other parameters between 0 and 1.
```

File: api/cpp/include/slint_color.h [hunk 6]
```diff
@@ -117,22 +130,40 @@ class Color
...
+    [[nodiscard]] static Color from_hsva(float h, float s, float v, float a)
+    {
+        Color ret;
+        ret.inner = cbindgen_private::types::slint_color_from_hsva(h, s, v, a);
+        return ret;
+    }
+
+    /// Convert this color to HSV
```
Comment (tronical) [inline api/cpp/include/slint_color.h:147]:
```suggestion
    /// Convert this color to the HSV color space.
    /// @returns a new HsvaColor.
```

File: docs/reference/src/language/builtins/namespaces.md [hunk 7]
```diff
@@ -21,6 +21,11 @@ The fourth value, if present, is an alpha value between 0 and 1.
 
 Unlike in CSS, the commas are mandatory.
 
+ - **`hsv(h: float, s: float, v: float) -> color`**, **`hsv(h: float, s: float, v: float, a: float) -> color`**
+
+Return a color computed from the HSV color space. The hue is between 0 and 360.
+The saturation, value, and optional alpha parameter are comprised between 0 and 1.
```
Comment (tronical) [inline docs/reference/src/language/builtins/namespaces.md:27]:
```suggestion
The saturation, value, and optional alpha parameter are expected to be within the range of 0 and 1.
```

File: docs/reference/src/language/syntax/types.md [hunk 8]
```diff
@@ -103,11 +103,15 @@ All colors and brushes define the following methods:
...
-
 -  **`with_alpha(alpha: float) -> brush`**
 
     Returns a new color with the alpha value set to `alpha` (between 0 and 1)
 
+- **`to-hsv()->{hue: float, saturation: float, value: float}`**
+
+    Returns a struct with the fields in the HSV colorspace.
+    Hue is between 0 and 360 while saturation and value are between 0 and 1.
+
```
Comment (tronical) [inline docs/reference/src/language/syntax/types.md:114]: Either this:

```suggestion
- **`to-hsv()->{hue: float, saturation: float, value: float, alpha: float}`**

    Converts this color to the HSV color space and returns a struct with the `hue`, `saturation`, `value`,
    and `alpha` fields. `hue` is between 0 and 360 while `saturation`, `value`, and `alpha` are between 0 and 1.

```

or

```suggestion
- **`to-hsv()->{hue: float, saturation: float, value: float}`**

    Converts this color to the HSV color space and returns a struct with the `hue`, `saturation`, and `value`
    fields. `hue` is between 0 and 360 while `saturation` and `value` are between 0 and 1.
    The Color's alpha value is discarded.
```

I think the former makes more sense though, as we ... don't really discard it.

File: internal/compiler/generator/rust.rs [hunk 9]
```diff
@@ -2539,6 +2540,17 @@ fn compile_builtin_function_call(
                 sp::Color::from_argb_u8(a, r, g, b)
             })
         }
+        BuiltinFunction::Hsv => {
+            let (h, s, v, a) =
+                (a.next().unwrap(), a.next().unwrap(), a.next().unwrap(), a.next().unwrap());
+            quote!({
+                let h: f32 = (#h as f32).max(0.).min(360.) as f32;
```
Comment (tronical) [inline internal/compiler/generator/rust.rs:2547]: Just curious... is there a specific reason to prefer `.max(0.).min(360.)` over `.clamp(0., 360.)`?

File: internal/core/graphics/color.rs [hunk 10]
```diff
@@ -145,6 +145,20 @@ impl Color {
         RgbaColor::from(*self)
     }
 
+    /// Converts this color to the HSV color space
```
Comment (tronical) [inline internal/core/graphics/color.rs:148]:
```suggestion
    /// Converts this color to the HSV color space.
```

File: internal/core/graphics/color.rs [hunk 11]
```diff
@@ -145,6 +145,20 @@ impl Color {
...
+    pub fn to_hsva(&self) -> HsvaColor {
+        let rgba: RgbaColor<f32> = (*self).into();
+        rgba.into()
+    }
+
+    /// Construct a color from the hue, saturation, and value HSV color parameters.
```
Comment (tronical) [inline internal/core/graphics/color.rs:154]:
```suggestion
    /// Construct a color from the hue, saturation, and value HSV color space parameters.
```

File: internal/core/graphics/color.rs [hunk 12]
```diff
@@ -145,6 +145,20 @@ impl Color {
...
+    ///
+    /// Hue is between 0 and 360, the others parameters between 0 and 1
```
Comment (tronical) [inline internal/core/graphics/color.rs:156]:
```suggestion
    /// Hue is between 0 and 360, the others parameters between 0 and 1.
```

File: internal/core/graphics/color.rs [hunk 13]
```diff
@@ -319,12 +333,20 @@ impl core::fmt::Display for Color {
...
 }
 
-#[derive(Debug, Clone, Copy, PartialEq)]
-struct HsvaColor {
-    h: f32,
-    s: f32,
-    v: f32,
-    alpha: f32,
+/// HsvaColor stores the hue, saturation, value and alpha components of a color
+/// as `f32 `.
```
Comment (tronical) [inline internal/core/graphics/color.rs:337]:
```suggestion
/// in the HSV color space as `f32 ` fields.
```

File: internal/core/graphics/color.rs [hunk 14]
```diff
@@ -319,12 +333,20 @@ impl core::fmt::Display for Color {
...
+/// This is merely a helper class for use with [`Color`].
```
Comment (tronical) [inline internal/core/graphics/color.rs:338]:
```suggestion
/// This is merely a helper struct for use with [`Color`].
```

File: internal/core/graphics/color.rs [hunk 15]
```diff
@@ -319,12 +333,20 @@ impl core::fmt::Display for Color {
...
+#[derive(Copy, Clone, PartialEq, PartialOrd, Debug, Default)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct HsvaColor {
+    /// The hue component in degrees between 0 and 360
```
Comment (tronical) [inline internal/core/graphics/color.rs:342]:
```suggestion
    /// The hue component in degrees between 0 and 360.
```

File: internal/core/graphics/color.rs [hunk 16]
```diff
@@ -319,12 +333,20 @@ impl core::fmt::Display for Color {
...
+    pub hue: f32,
+    /// The saturation component, between 0 and 1
```
Comment (tronical) [inline internal/core/graphics/color.rs:344]:
```suggestion
    /// The saturation component, between 0 and 1.
```

File: internal/core/graphics/color.rs [hunk 17]
```diff
@@ -319,12 +333,20 @@ impl core::fmt::Display for Color {
...
+    pub saturation: f32,
+    /// The value component, between 0 and 1
```
Comment (tronical) [inline internal/core/graphics/color.rs:346]:
```suggestion
    /// The value component, between 0 and 1.
```

File: internal/core/graphics/color.rs [hunk 18]
```diff
@@ -319,12 +333,20 @@ impl core::fmt::Display for Color {
...
+    pub value: f32,
+    /// The alpha component, between 0 and 1
```
Comment (tronical) [inline internal/core/graphics/color.rs:348]:
```suggestion
    /// The alpha component, between 0 and 1.
```

File: internal/compiler/generator/rust.rs (hunk 9 above)
Comment (ogoffart) [reply internal/compiler/generator/rust.rs:2547]: I don't think there is a reason.

```suggestion
                let h: f32 = (#h as f32).clamp(0., 360.) as f32;
```

**Pull request merged**